File: BI-TJV/semestral/rest-server/src/main/java/wrzecond/controller/TjvController.java
package wrzecond.controller;

import lombok.AllArgsConstructor;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.http.HttpStatus;
import org.springframework.orm.jpa.JpaObjectRetrievalFailureException;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import wrzecond.ICreateDTO;
import wrzecond.IReadDTO;
import wrzecond.entity.TjvEmployee;
import wrzecond.entity.TjvEntity;
import wrzecond.service.TjvEmployeeService;
import wrzecond.service.TjvService;

import javax.persistence.EntityNotFoundException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.lang.annotation.*;
import java.util.List;
import java.util.function.Function;

@AllArgsConstructor
abstract class TjvController<T extends TjvEntity, R extends IReadDTO, C extends ICreateDTO> {

    // === PARAMETERS ===
    private final TjvService<T, R, C> service;
    private final TjvEmployeeService employeeService;
    protected abstract String getCreateLocationPath ();

    // === AUTHENTICATION ===
    private void authenticate (HttpServletRequest request, Function<Visibility, VisibilitySettings> function) {
        Visibility annotation = getClass().getAnnotation(Visibility.class);
        authenticate(request, function.apply(annotation));
    }
    protected TjvEmployee authenticate (HttpServletRequest request, VisibilitySettings settings) {
        // Init properties from request
        String username = request.getHeader("Username");
        String password = request.getHeader("Password");
        TjvEmployee employee = employeeService.getByUsernamePassword(username, password);

        // See authorization status
        switch (settings) {
            case ALL:
                break;
            case LOGGED:
                // Not logged in
                if (employee == null)
                    throw new ResponseStatusException(HttpStatus.UNAUTHORIZED);
                break;
            case ADMIN:
                // Not logged in
                if (employee == null)
                    throw new ResponseStatusException(HttpStatus.UNAUTHORIZED);
                // Not sufficient permissions
                if (!employee.getAdmin())
                    throw new ResponseStatusException(HttpStatus.FORBIDDEN);
                break;
            default:
            case NONE:
                throw new ResponseStatusException(HttpStatus.METHOD_NOT_ALLOWED);
        }

        return employee;
    }

    @GetMapping
    List<R> all (HttpServletRequest request) {
        authenticate(request, Visibility::findAll);
        return service.findAll();
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    void create (@RequestBody C table, HttpServletRequest request, HttpServletResponse response) {
        authenticate(request, Visibility::create);
        try {
            R dto = service.create(table);
            response.addHeader("Location", String.format(getCreateLocationPath(), dto.getId()));
        }
        catch (DataIntegrityViolationException exception) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
        }
    }

    @PatchMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void update (@PathVariable int id, @RequestBody C table, HttpServletRequest request) {
        authenticate(request, Visibility::update);
        try {
            service.update(id, table);
        }
        catch (EntityNotFoundException | JpaObjectRetrievalFailureException exc) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
        catch (DataIntegrityViolationException exception) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
        }
    }

    @GetMapping("/{id}")
    R getByID (@PathVariable int id, HttpServletRequest request) {
        authenticate(request, Visibility::getByID);
        try {
            return service.getByID(id);
        }
        catch ( EntityNotFoundException exc ) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void delete (@PathVariable int id, HttpServletRequest request) {
        authenticate(request, Visibility::delete);
        try {
            service.delete(id);
        }
        catch ( EmptyResultDataAccessException exception ) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    protected @interface Visibility {
        VisibilitySettings findAll() default VisibilitySettings.ALL;
        VisibilitySettings getByID() default VisibilitySettings.ALL;
        
        VisibilitySettings create() default VisibilitySettings.ADMIN;
        VisibilitySettings delete() default VisibilitySettings.ADMIN;
        VisibilitySettings update() default VisibilitySettings.ADMIN;
    }
    protected enum VisibilitySettings {
        NONE,
        ADMIN,
        LOGGED,
        ALL
    }

}

File: BI-KOT/semestral/rest-server/src/main/kotlin/wrzecond/controller/TjvController.kt
package wrzecond.controller

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import wrzecond.ICreateDTO
import wrzecond.IReadDTO
import wrzecond.IUpdateDTO
import wrzecond.entity.TjvEmployee
import wrzecond.entity.TjvEntity
import wrzecond.service.TjvEmployeeService
import wrzecond.service.TjvService
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import kotlin.reflect.full.findAnnotation

abstract class TjvController<T: TjvEntity, R: IReadDTO, C: ICreateDTO, U: IUpdateDTO>
    (protected open val service: TjvService<T, R, C, U>, protected val employeeService: TjvEmployeeService) {

    protected abstract val createLocationPath: String

    // === AUTHENTICATION ===
    private fun <A> authenticate (request: HttpServletRequest, function: (Visibility) -> VisibilitySettings, action: (TjvEmployee?) -> A) : A {
        val annotation = this::class.findAnnotation<Visibility>()
        return authenticate(request, function(annotation!!), action)
    }
    protected fun <A> authenticate (request: HttpServletRequest, settings: VisibilitySettings, action: (TjvEmployee?) -> A) : A {
        // Load data from header
        val username = request.getHeader("Username") ?: ""
        val password = request.getHeader("Password") ?: ""

        // Try to auth, exception = not authed
        val employee = try { employeeService.getByUsernamePassword(username, password) }
        catch (e: ResponseStatusException) { null }

        // Throw UNAUTHORIZED/FORBIDDEN if we need to be authed
        when (settings) {
            VisibilitySettings.ALL    -> {}
            VisibilitySettings.LOGGED -> if (employee == null) throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
            VisibilitySettings.ADMIN  -> {
                if (employee == null)
                    throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
                if (!employee.admin)
                    throw ResponseStatusException(HttpStatus.FORBIDDEN)
            }
            VisibilitySettings.NONE   -> throw ResponseStatusException(HttpStatus.METHOD_NOT_ALLOWED)
        }

        // Perform requested action
        return action(employee)
    }

    // Helper auth annotation class
    annotation class Visibility (
        val findAll: VisibilitySettings = VisibilitySettings.ALL,
        val getByID: VisibilitySettings = VisibilitySettings.ALL,
        val create : VisibilitySettings = VisibilitySettings.ADMIN,
        val delete : VisibilitySettings = VisibilitySettings.ADMIN,
        val update : VisibilitySettings = VisibilitySettings.ADMIN
    )
    enum class VisibilitySettings { NONE, ADMIN, LOGGED, ALL }
    // === AUTHENTICATION ===

    @GetMapping
    open fun all (request: HttpServletRequest) = authenticate(request, {it.findAll}) {
        service.findAll()
    }

    @GetMapping("/{id}")
    fun getByID (@PathVariable id: Int, request: HttpServletRequest) = authenticate(request, {it.getByID}) {
        service.getByID(id)
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create (@RequestBody createDTO: C, request: HttpServletRequest, response: HttpServletResponse)
    = authenticate(request, {it.create}) {
        val dto = service.create(createDTO)[0]
        response.addHeader("Location", String.format(createLocationPath, dto.id))
    }

    @PatchMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun create (@PathVariable id: Int, @RequestBody updateDTO: U, request: HttpServletRequest)
    = authenticate(request, {it.update}) {
        service.update(id, updateDTO)
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete (@PathVariable id: Int, request: HttpServletRequest)
    = authenticate(request, {it.delete}) {
        service.delete(id)
    }

}